File: src/test/scala/parse/dont/validate/NIRProperties.scala
package parse.dont.validate

import org.scalacheck.Prop.forAll
import org.scalacheck.{Arbitrary, Gen, Properties}
import org.scalatest.EitherValues
import parse.dont.validate.Generators.validNIRGenerator
import parse.dont.validate.NIRDomain.*
import parse.dont.validate.NIRDomain.NIR.parseNIR

object NIRProperties extends Properties("NIR") with EitherValues {
  implicit def nirGenerator: Arbitrary[NIR] = validNIRGenerator()

  property("isomorphic") = forAll { (nir: NIR) =>
    val s = nir.toString
    parseNIR(s).value == nir
  }
}

File: solution/java/src/test/java/parse/dont/validate/NIRProperties.java
package parse.dont.validate;

import io.vavr.test.Arbitrary;
import io.vavr.test.Gen;
import io.vavr.test.Property;
import org.junit.jupiter.api.Test;

class NIRProperties {
    private final Gen<Sex> sexGenerator = Gen.choose(Sex.values());
    private final Arbitrary<NIR> validNIR =
            sexGenerator.map(NIR::new)
                    .arbitrary();

    @Test
    void roundTrip() {
        Property.def("parseNIR(nir.ToString()) == nir")
                .forAll(validNIR)
                .suchThat(nir -> NIR.parseNIR(nir.toString()).contains(nir))
                .check()
                .assertIsSatisfied();
    }
}
